src/style.js

```diff
@@ -1,35 +1,29 @@
-import { useLayoutEffect } from 'react'
-import StyleSheetRegistry from './stylesheet-registry'
-
-const styleSheetRegistry = new StyleSheetRegistry()
+import { useLayoutEffect, useContext } from 'react'
+import { StyleSheetContext } from './stylesheet-registry'
 
 export default function JSXStyle(props) {
+  const registry = useContext(StyleSheetContext)
   if (typeof window === 'undefined') {
-    styleSheetRegistry.add(props)
+    registry.add(props)
     return null
   }
   useLayoutEffect(() => {
-    styleSheetRegistry.add(props)
+    registry.add(props)
     return () => {
-      styleSheetRegistry.remove(props)
+      registry.remove(props)
     }
     // props.children can be string[], will be striped since id is identical
   }, [props.id, String(props.dynamic)])
   return null
 }
 
 JSXStyle.dynamic = info => {
+  const registry = useContext(StyleSheetContext)
```

Comment:
I am afraid that when you have an early return in a component this will result in an error. See https://codesandbox.io/s/ecstatic-darkness-8yd0e?file=/src/App.js

<img width="1163" alt="Screen Shot 2021-08-11 at 9 48 05 AM" src="https://user-images.githubusercontent.com/711311/128990752-d84daa49-3518-41e7-b4e2-683075bf907a.png">

Reply:
oh nevermind probably that's never going to be the case because the Babel plugin adds the `JSXStyle.dynamic` to every JSX element

Reply:
`registry.computeId` and `registry.computeSelector` are pure functions so you could move them outside of the `StyleSheetRegistry` class and have them share a global cache. This would allow you to remove the context dependency from `JSXStyle.dynamic`

Reply:
Nice callcout! I think the cache of `registry.computeId` and `registry.computeSelector` should live per request like the style registry context, to avoid conflicts in concurrent mode? Might be bit hard to do if they don't consume cache or registry from context?

Reply:
The cache can be shared because it is there just to speed things up a bit. There won't be conflicts

https://github.com/vercel/styled-jsx/blob/d591ce27c9ae4d41d7833551e5a153324cdd69ca/src/stylesheet-registry.js#L134-L150

You can turn this into an IIFE, or just define the cache as a local variable in the module

Reply:
I wonder if we can rename `JSXStyle.dynamic` to `JSXStyle.useDynamic` since it's technically a hook now

Reply:
I am suggesting to remove the dependency from context because it is not needed to compute the ids. We just need to move `computeId` and `computeSelector` outside of the `StyleSheetRegistry` class and export them.

Reply:
I mean we don't need those factories `createComputeId` and `createComputeSelector` and one version of the resulting function per `StyleSheetRegistry` instance. The two caches can be shared